File: python-ai/utils/fileUtils.py
"""
File Utilities for Python AI Service
"""

import os
import hashlib
import re
from pathlib import Path
from typing import List, Set, Optional
from config.projectPaths import SUPPORTED_EXTENSIONS, EXCLUDED_DIRS, EXCLUDED_PATTERNS
from utils.logger import get_logger

logger = get_logger(__name__)

def calculate_file_hash(file_path: str) -> Optional[str]:
    """Calculate SHA256 hash of file content"""
    try:
        with open(file_path, 'r', encoding='utf-8', errors='ignore') as f:
            content = f.read()
        return hashlib.sha256(content.encode()).hexdigest()
    except Exception as e:
        logger.error(f"Error calculating file hash: {file_path}, {str(e)}")
        return None


def should_process_file(file_path: str) -> bool:
    """Check if file should be processed"""
    filename = os.path.basename(file_path)
    ext = os.path.splitext(file_path)[1]

    # Check excluded patterns
    for pattern in EXCLUDED_PATTERNS:
        if re.search(pattern, filename):
            return False

    # Check supported extensions
    if ext in SUPPORTED_EXTENSIONS or filename in SUPPORTED_EXTENSIONS:
        return True

    return False


def should_exclude_dir(dir_path: str) -> bool:
    """Check if directory should be excluded"""
    dirname = os.path.basename(dir_path)
    return dirname in EXCLUDED_DIRS


def scan_directory(dir_path: str, max_depth: int = 10, current_depth: int = 0) -> List[str]:
    """Recursively scan directory for files"""
    files = []

    if current_depth >= max_depth:
        return files

    try:
        for entry in os.listdir(dir_path):
            full_path = os.path.join(dir_path, entry)

            if os.path.isdir(full_path):
                if not should_exclude_dir(full_path):
                    files.extend(scan_directory(full_path, max_depth, current_depth + 1))
            elif os.path.isfile(full_path):
                if should_process_file(full_path):
                    files.append(full_path)
    except Exception as e:
        logger.warning(f"Error scanning directory: {dir_path}, {str(e)}")

    return files


def read_file_content(file_path: str) -> Optional[str]:
    """Read file content with encoding detection"""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            return f.read()
    except UnicodeDecodeError:
        try:
            with open(file_path, 'r', encoding='latin-1') as f:
                return f.read()
        except Exception as e:
            logger.error(f"Error reading file: {file_path}, {str(e)}")
            return None
    except Exception as e:
        logger.error(f"Error reading file: {file_path}, {str(e)}")
        return None


def get_file_metadata(file_path: str) -> dict:
    """Get file metadata"""
    try:
        stat = os.stat(file_path)
        filename = os.path.basename(file_path)
        ext = os.path.splitext(file_path)[1]

        # Determine language
        language_map = {
            '.js': 'javascript',
            '.jsx': 'javascript',
            '.ts': 'typescript',
            '.tsx': 'typescript',
            '.py': 'python',
            '.java': 'java',
            '.kt': 'kotlin',
            '.dart': 'dart',
            '.md': 'markdown',
            '.json': 'json',
            '.yaml': 'yaml',
            '.yml': 'yaml',
        }
        language = language_map.get(ext, 'other')

        # Determine file type
        file_type = 'other'
        if filename in ['README.md', '.env', '.env.example', 'package.json', 'pubspec.yaml', 'requirements.txt']:
            file_type = 'config'
        elif ext in ['.md', '.txt']:
            file_type = 'documentation'
        elif ext in ['.js', '.jsx', '.ts', '.tsx', '.py', '.java', '.kt', '.dart']:
            file_type = 'code'
        elif ext in ['.json', '.yaml', '.yml']:
            file_type = 'data'

        return {
            'filename': filename,
            'language': language,
            'file_type': file_type,
            'file_size': stat.st_size,
            'created_at': stat.st_ctime,
            'modified_at': stat.st_mtime,
        }
    except Exception as e:
        logger.error(f"Error getting file metadata: {file_path}, {str(e)}")
        return None


def detect_project_type(project_path: str) -> str:
    """Detect project type from directory"""
    try:
        files = os.listdir(project_path)

        if 'package.json' in files:
            return 'nodejs'
        if 'requirements.txt' in files or 'setup.py' in files:
            return 'python'
        if 'pom.xml' in files or 'build.gradle' in files:
            return 'java'
        if 'pubspec.yaml' in files:
            return 'flutter'

        return 'other'
    except Exception as e:
        logger.error(f"Error detecting project type: {project_path}, {str(e)}")
        return 'other'


def get_project_name(project_path: str) -> str:
    """Get project name from directory"""
    return os.path.basename(project_path)

File: backend/ai/utils/fileUtils.js
/**
 * File Utilities for AI System
 * Handles file operations, hashing, and content extraction
 */

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const { SUPPORTED_EXTENSIONS, EXCLUDED_DIRS, EXCLUDED_PATTERNS } = require('../config/projectPaths');
const { resolveModuleLabel, isAllowedTrainingPath } = require('../config/proposalForgeModules');
const AILogger = require('./logger');

const logger = new AILogger('FileUtils');

/**
 * Calculate file hash for change detection
 */
function calculateFileHash(filePath) {
  try {
    const content = fs.readFileSync(filePath, 'utf-8');
    return crypto.createHash('sha256').update(content).digest('hex');
  } catch (err) {
    logger.error('Error calculating file hash', { filePath, error: err.message });
    return null;
  }
}

/**
 * Check if file should be processed
 */
function shouldProcessFile(filePath) {
  const filename = path.basename(filePath);
  const ext = path.extname(filePath);

  if (!isAllowedTrainingPath(filePath)) {
    return false;
  }

  for (const pattern of EXCLUDED_PATTERNS) {
    if (pattern.test(filename)) {
      return false;
    }
  }

  // Check supported extensions
  const isSupported = SUPPORTED_EXTENSIONS.some(supportedExt => {
    if (supportedExt.startsWith('.')) {
      return ext === supportedExt || filename === supportedExt;
    }
    return filename === supportedExt;
  });

  return isSupported;
}

/**
 * Check if directory should be excluded
 */
function shouldExcludeDir(dirPath) {
  const dirname = path.basename(dirPath);
  return EXCLUDED_DIRS.includes(dirname);
}

/**
 * Recursively scan directory for files
 */
function scanDirectory(dirPath, maxDepth = 10, currentDepth = 0) {
  const files = [];

  if (currentDepth >= maxDepth) {
    return files;
  }

  try {
    const entries = fs.readdirSync(dirPath, { withFileTypes: true });

    for (const entry of entries) {
      const fullPath = path.join(dirPath, entry.name);

      if (entry.isDirectory()) {
        if (!shouldExcludeDir(fullPath)) {
          files.push(...scanDirectory(fullPath, maxDepth, currentDepth + 1));
        }
      } else if (entry.isFile()) {
        if (shouldProcessFile(fullPath)) {
          files.push(fullPath);
        }
      }
    }
  } catch (err) {
    logger.warn('Error scanning directory', { dirPath, error: err.message });
  }

  return files;
}

/**
 * Read file content with encoding detection
 */
function readFileContent(filePath) {
  try {
    // Try UTF-8 first
    const content = fs.readFileSync(filePath, 'utf-8');
    return content;
  } catch (err) {
    try {
      // Fallback to latin1
      const content = fs.readFileSync(filePath, 'latin1');
      return content;
    } catch (fallbackErr) {
      logger.error('Error reading file content', { filePath, error: err.message });
      return null;
    }
  }
}

/**
 * Get file metadata
 */
function getFileMetadata(filePath) {
  try {
    const stats = fs.statSync(filePath);
    const ext = path.extname(filePath);
    const filename = path.basename(filePath);

    // Determine language
    let language = 'other';
    if (['.js', '.jsx'].includes(ext)) language = 'javascript';
    else if (['.ts', '.tsx'].includes(ext)) language = 'typescript';
    else if (['.py'].includes(ext)) language = 'python';
    else if (['.java'].includes(ext)) language = 'java';
    else if (['.kt'].includes(ext)) language = 'kotlin';
    else if (['.dart'].includes(ext)) language = 'dart';
    else if (['.md'].includes(ext)) language = 'markdown';
    else if (['.json'].includes(ext)) language = 'json';
    else if (['.yaml', '.yml'].includes(ext)) language = 'yaml';

    // Determine file type
    let fileType = 'other';
    if (['README.md', '.env', '.env.example', 'package.json', 'pubspec.yaml', 'requirements.txt'].includes(filename)) {
      fileType = 'config';
    } else if (['.md', '.txt'].includes(ext)) {
      fileType = 'documentation';
    } else if (['.js', '.jsx', '.ts', '.tsx', '.py', '.java', '.kt', '.dart'].includes(ext)) {
      fileType = 'code';
    } else if (['.json', '.yaml', '.yml'].includes(ext)) {
      fileType = 'data';
    }

    return {
      filename,
      language,
      fileType,
      fileSize: stats.size,
      createdAt: stats.birthtime,
      modifiedAt: stats.mtime,
    };
  } catch (err) {
    logger.error('Error getting file metadata', { filePath, error: err.message });
    return null;
  }
}

/**
 * Detect project type from directory
 */
function detectProjectType(projectPath) {
  const files = fs.readdirSync(projectPath, { withFileTypes: true });
  const fileNames = files.map(f => f.name);

  if (fileNames.includes('package.json')) {
    const packageJson = JSON.parse(fs.readFileSync(path.join(projectPath, 'package.json'), 'utf-8'));
    if (packageJson.dependencies?.react || packageJson.devDependencies?.react) {
      return 'react';
    }
    return 'nodejs';
  }

  if (fileNames.includes('requirements.txt') || fileNames.includes('setup.py')) {
    return 'python';
  }

  if (fileNames.includes('pom.xml') || fileNames.includes('build.gradle')) {
    return 'java';
  }

  return 'other';
}

/**
 * Get project name from directory
 */
function getProjectName(projectPath) {
  return resolveModuleLabel(projectPath);
}

module.exports = {
  calculateFileHash,
  shouldProcessFile,
  shouldExcludeDir,
  scanDirectory,
  readFileContent,
  getFileMetadata,
  detectProjectType,
  getProjectName,
};
